### DISCUSSION TITLE
Color Issue in AEM assets

### DISCUSSION THREAD

**Discussion by HardikCharadva** [ImageMagick/ImageMagick#3440] [Help]
Hi Team,
CMYK-based files in the EPS/PSD formats render incorrect shades of colors for the system created thumbnail renditions using Imagemagick.

Can you please guide us how to preserve the colors while this conversion? What parameters we should add to command?

**Command :**
convert ${filename} -define jpeg:size=319x319 -thumbnail "319x319>" -background transparent -gravity center -extent 319x319 -write png:cq5dam.thumbnail.319.319.png -thumbnail "140x100>" -background transparent -gravity center -extent 140x100 -write cq5dam.thumbnail.140.100.png -thumbnail "48x48>" -background transparent -gravity center -extent 48x48 cq5dam.thumbnail.48.48.png

Thanks
Hardik

**Comment by snibgo** [comment]
What version of IM, on what platform?

You have "-define jpeg:size=319x319" but don't write to JPEG, so that can be removed.

Your outputs are PNG format, which cannot record CMYK, so IM will automatically convert to sRGB. One solution is to write to a format that can record CMYK, such as TIFF.

If your input has an embedded ICC profile, a more accurate conversion to sRGB would be with the "-profile" operator like this:
```
convert ${filename} -profile sRGB.icc ... etc
```

**Reply by HardikCharadva** [reply]
Hi @snibgo - 
 Version: Imagemagick 7.0.8-68 Q16 x86_64 2019-10-21 https://imagemagick.org

We are using Imagemagick for generating assets renditions on Adobe experience manager. We have requirement of generating png renditions. (i.e. Dimensions : 319*319, 140*100, 48*48)

To create the web-enabled rendition using Imagemagick:
convert ${filename} -define jpeg:size=1280x1280 -thumbnail "1280x1280>" cq5dam.web.1280.1280.jpeg

We use above command which gives proper CMYK colorspace image. But for png it's converting to sRGB.

So to avoid color issue , What's the best practice to generate above renditions using Imagemagick. ?

Thanks
Hardik

**Comment by snibgo** [comment]
For IM v7, I suggest using "magick", not "convert" or "magick convert".

Does your input file have an embedded ICC profile? Please show the text output from "magick ${filename} -verbose info:".

**Comment by HardikCharadva** [comment]
Image: 0-34000-37104-4.eps
  Format: EPT (Encapsulated PostScript with TIFF preview)
  Class: DirectClass
  Geometry: 87x62+0+0
  Resolution: 72x72
  Print size: 1.20833x0.861111
  Units: Undefined
  Colorspace: CMYK
  Type: ColorSeparation
  Endianess: Undefined
  Depth: 16/8-bit
  Channel depth:
    Cyan: 1-bit
    Magenta: 1-bit
    Yellow: 1-bit
    Black: 8-bit
  Channel statistics:
    Pixels: 5394
    Cyan:
      min: 0  (0)
      max: 0 (0)
      mean: 0 (0)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 4.82164e-05
    Magenta:
      min: 0  (0)
      max: 0 (0)
      mean: 0 (0)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 4.82164e-05
    Yellow:
      min: 0  (0)
      max: 0 (0)
      mean: 0 (0)
      standard deviation: 0 (0)
      kurtosis: -3
      skewness: 0
      entropy: 4.82164e-05
    Black:
      min: 0  (0)
      max: 65535 (1)
      mean: 27544.4 (0.4203)
      standard deviation: 25667.9 (0.391667)
      kurtosis: -1.50981
      skewness: 0.190247
      entropy: 0.607214
  Image statistics:
    Overall:
      min: 0  (0)
      max: 65535 (1)
      mean: 6886.09 (0.105075)
      standard deviation: 6416.97 (0.0979167)
      kurtosis: 4.39853
      skewness: 2.4117
      entropy: 0.15184
  Total ink density: 100%
  Colors: 16
  Histogram:
      2159: (    0,    0,    0,    0) #0000000000000000 cmyk(0,0,0,0)
       993: (    0,    0,    0,48059) #000000000000BBBB cmyk(0,0,0,187)
       989: (    0,    0,    0,65535) #000000000000FFFF cmyk(0,0,0,255)
       517: (    0,    0,    0,34952) #0000000000008888 cmyk(0,0,0,136)
       365: (    0,    0,    0,29812) #0000000000007474 cmyk(0,0,0,116)
       208: (    0,    0,    0,17219) #0000000000004343 cmyk(0,0,0,67)
        53: (    0,    0,    0,25957) #0000000000006565 cmyk(0,0,0,101)
        34: (    0,    0,    0,30326) #0000000000007676 cmyk(0,0,0,118)
        25: (    0,    0,    0,21588) #0000000000005454 cmyk(0,0,0,84)
        16: (    0,    0,    0, 8481) #0000000000002121 cmyk(0,0,0,33)
        11: (    0,    0,    0, 4112) #0000000000001010 cmyk(0,0,0,16)
        10: (    0,    0,    0,15934) #0000000000003E3E cmyk(0,0,0,62)
         9: (    0,    0,    0,12850) #0000000000003232 cmyk(0,0,0,50)
         2: (    0,    0,    0,23387) #0000000000005B5B cmyk(0,0,0,91)
         2: (    0,    0,    0, 7967) #0000000000001F1F cmyk(0,0,0,31)
         1: (    0,    0,    0,39321) #0000000000009999 cmyk(0,0,0,153)
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 87x62+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Undefined
  Orientation: Undefined
  Properties:
    date:create: 2021-03-24T12:38:15+00:00
    date:modify: 2021-03-24T12:38:15+00:00
    pdf:SpotColor-0: DocumentCustomColors:
    pdf:SpotColor-1: CMYKCustomColor:
    ps:HiResBoundingBox: 87x62+0+0
    ps:Level: !PS-Adobe-3.0 EPSF-3.0
    signature: 9c7103b9aca86c6726d5bf4bd69de746629bcd6a782b264eb04612ee073aa933
  Tainted: False
  Filesize: 21675B
  Number pixels: 5394
  Pixels per second: 3.06999MP
  User time: 0.000u
  Elapsed time: 0:01.001
  Version: ImageMagick 7.0.8-68 Q16 x86_64 2019-10-21 https://imagemagick.org

**Comment by snibgo** [comment]
So your input is a CMYK EPS with no embedded profile. For all pixels, C=M=Y=zero. The K (black) channel varies. But you get "incorrect shades of colors".

I don't understand. Do you mean you get incorrect shades of gray? Are they too light, or too dark, or what?

**Reply by HardikCharadva** [reply]
@snibgo  Sir, Not sure about particular asset - but in general what command we should use that's I am searching here...might be profile attached/might not, but command should convert the EPS/PSD with right color renditions.

While we were using Imagemagick we had a color issue, colors were not correct. So can you provide us the right command which we should use to generate good renditions of EPS/PSD assets?

**Comment by snibgo** [comment]
If a CMYK immage has no embedded ICC profile, and you don't know what profile should be embedded, then any conversion to sRGB has to guess what profile to use.

**Reply by HardikCharadva** [reply]
So sir, you mean to say if we attach CMYK.icc profile to EPS/PSD files , then it will pick right profile and covert the images accurately?

**Comment by snibgo** [comment]
Yes. For example, we might guess the correct profile is USWebCoatedSWOP:
```
magick in.eps -profile USWebCoatedSWOP.icc -profile sRGB.icc out.png
```

Or we can tell Ghostscript to do the conversion to sRGB:
```
magick -colorspace sRGB in.eps out.png
```

**Reply by HardikCharadva** [reply]
@snibgo It's final that it generate the output file in sRGB. So sir, if input file has the profile attached like you mentioned, CMYK.icc, and that file if we process - whatever renditons it will generate, will it accurate in terms of color? like our goal is input original file and renditions should be accurate colors.

**Comment by snibgo** [comment]
The gamuts of sRGB and any give CMYK colorspace, such as USWebCoatedSWOP, are different. This means that some colours that are represented in USWebCoatedSWOP cannot be represented in sRGB, and vice versa.

So the transformation from one colorspace to another will involve changing colours to ensure they all fit inside the new gamut. The method used for this is influenced by the "-intent" setting.

**Reply by hdc11071994** [reply]
Hi Sir,
Can you help me to provide command in one line for below renditions generation of eps asset?

Dimensions : 319x319, 140x100, 48x48 , Web rendition : 1280x1280 

We wanted to generate png for first 3 dimensions and web rendition needs to create with colorspace sRGB and of jpeg extension.

We would like to put this command in adobe experience manager workflow command line step, so when workflow run it will generate all 4 renditions properly.

Thanks in advance.

**Comment by snibgo** [comment]
I know nothing about "adobe experience manager". I assume you want the outputs to be exactly those dimensions, and you want bash syntax:
```
magick \
  -colorspace sRGB in.eps \
  \( +clone -resize 319x319! -write x1.png +delete \) \
  \( +clone -resize 140x100! -write x2.png +delete \) \
  \( +clone -resize 48x48! -write x3.png +delete \) \
  -resize 1280x1280! out.jpg
```

**Reply by hdc11071994** [reply]
Thank you @snibgo , What parameter we can add if renditions looks blurry? Can we use thumbnail instead of resize?

**Reply by hdc11071994** [reply]
Hi @snibgo 
[Assets.zip](https://github.com/ImageMagick/ImageMagick/files/6465986/Assets.zip)

We use below command to generate renditions :
```
convert -define jpeg:size=319x319 ${filename} -thumbnail 319x319 cq5dam.thumbnail.319.319.png
convert -define jpeg:size=140x100 ${filename} -thumbnail 140x100 cq5dam.thumbnail.140.100.png
convert -define jpeg:size=48x48 ${filename} -thumbnail 48x48 cq5dam.thumbnail.48.48.png
convert -define jpeg:size=1280x1280 ${filename} -thumbnail 1280x1280 -colorspace sRGB cq5dam.web.1280.1280.jpeg
```

But I can see 1280 renditions generated with black background, can you help me why ?

Thanks
Hardik

**Comment by snibgo** [comment]
Internally, "-thumbnail" does a "-resize", possibly after a "-sample". If this involves "-sample", the process will be faster but lower quality than just "-resize".

Your input eps files are vector. If you rasterize these at a low density and enlarge, the results will look horrible. Results will look better if you rasterize at a high density and shrink the result.

```
f:\web\im>%IMG7%magick circle.eps info:
circle.eps EPS 540x720 540x720+0+0 16-bit sRGB 14111B 0.063u 0:00.041
```
If we don't use "-density", the result is 540x720, so your "-thumbnail 1280x1280" will enlarge.

```
f:\web\im>%IMG7%magick -density 300 circle.eps info:
circle.eps EPS 2250x3000 2250x3000+0+0 16-bit sRGB 82939B 1.828u 0:00.507
```
Using "-density 300", the result is 2250x3000, so your "-thumbnail 1280x1280" will shrink, giving better quality.

As far as I know, "-define jpeg:size" is relevant only when reading JPEG files, so this has no effect for you.
